### PR TITLE
Refactored Action Management Functions for a Sprint 3 Basic Prototype

### PR DESCRIPTION
**Pull request by MTUC08**
Issue #115 completed. Refactored Action Kind functions to return strings to CLI. The functions now check for whether the action type given is the correct kind, and whether it can be used on an item. If checks pass, the function will return a string to CLI saying the action has been requested. If any check fails, the function will return a tailored error message, also in the form of a string. 

Note: After this PR to actions/dev, we need to wait on game-state to PR their updated items.h/c to master because our functions depend on a function they just added (issue #172), and will need that to be in place in order for the build to be successful.

### CODE REVIEW

File: src/action_management/src/actionmanagement.c [hunk 1]
```diff
@@ -3,7 +3,8 @@
 #include <stdio.h>
 #include "actionmanagement.h"
 #include "common.h"
-
+#include "string.h"
```
Comment (lucyyutingli) [inline src/action_management/src/actionmanagement.c]: I believe this should be <string.h>

File: src/action_management/src/actionmanagement.c [hunk 2]
```diff
@@ -50,58 +51,103 @@ void action_type_free(action_type_t *a)
...
+    assert(g->curr_player); // needed for sprint 4
     assert(a);
     assert(i);
+    char *ret_string = malloc(100); // buffer
+    ret_string[0] = '\0';
+    // checks if the action type is the correct kind
     if (a->kind != ITEM) {
-        fprintf(stderr, "The action type provided is not of the correct kind.\n");
-        return FAILURE;
+        strcat(ret_string, "The action type provided is not the right kind");
```
Comment (lucyyutingli) [inline src/action_management/src/actionmanagement.c]: I actually think the string "The action type provided is not of the correct kind" sounds a little better, but that could just be my opinion. I'll see what @ellychoi thinks.

File: src/action_management/src/actionmanagement.c [hunk 3]
```diff
@@ -50,58 +51,103 @@ void action_type_free(action_type_t *a)
...
     assert(g);
     assert(g->curr_player);
     assert(a);
+    char *ret_string = malloc(100); // buffer
+    ret_string[0] = '\0';
+    // checks if the action type is the correct kind
     if (a->kind != PATH) {
-        fprintf(stderr, "The action type provided is not of the correct kind.\n");
-        return FAILURE;
+        strcat(ret_string, "The action type provided is not the right kind");
```
Comment (lucyyutingli) [inline src/action_management/src/actionmanagement.c]: Same thing as I said earlier, wanna see @ellychoi 's opinion

File: src/action_management/src/actionmanagement.c [hunk 4]
```diff
@@ -50,58 +51,103 @@ void action_type_free(action_type_t *a)
...
+
+/* KIND 3
+ * See actionmanagement.h */
+char *do_item_item_action(game_t *g, action_type_t *a,
                         item_t *direct, item_t *indirect)
 {
     assert(g);
     assert(g->curr_player);
     assert(a);
+    assert(g->curr_player); // needed for sprint 4
```
Comment (lucyyutingli) [inline src/action_management/src/actionmanagement.c]: I see this twice, once on line 122, and once on line 124, is there supposed to be two?

File: src/action_management/src/actionmanagement.c [hunk 5]
```diff
@@ -50,58 +51,103 @@ void action_type_free(action_type_t *a)
...
     assert(direct);
     assert(indirect);
+    char *ret_string = malloc(100); // buffer
+    ret_string[0] = '\0';
+    // checks if the action type is the correct kind
     if (a->kind != ITEM_ITEM) {
-        fprintf(stderr, "The action type provided is not of the correct kind.\n");
-        return FAILURE;
+        strcat(ret_string, "The action type provided is not the right kind");
```
Comment (lucyyutingli) [inline src/action_management/src/actionmanagement.c]: Same as I said earlier, want @ellychoi 's opinion

File: src/action_management/src/actionmanagement.c [hunk 6]
```diff
@@ -50,58 +51,103 @@ void action_type_free(action_type_t *a)
...
+ * See actionmanagement.h */
+char *do_item_action(game_t *g, action_type_t *a, item_t *i)
 {
     // a couple confirmation checks
     assert(g);
-    assert(g->curr_player);
+    assert(g->curr_player); // needed for sprint 4
     assert(a);
     assert(i);
+    char *ret_string = malloc(100); // buffer
```
Comment (ellychoi) [inline src/action_management/src/actionmanagement.c]: When you say buffer, do you mean that this is placeholder memory and we will figure out another way to allocated memory with the sentences created?

File: src/action_management/src/actionmanagement.c [hunk 7]
```diff
@@ -50,58 +51,103 @@ void action_type_free(action_type_t *a)
...
+        strcat(ret_string, "The action type provided is not the right kind");
+        return ret_string;
     }
-    // TODO: implement the rest of this function, using game_state funcs
-    printf("Performed action %s on item %s", a->c_name, i->item_id);
-    return SUCCESS;
+    // checks if the action can be used on the item
+    int allowed = allowed_action(i, a->c_name);
+    if (allowed != SUCCESS) {
+        strcat(ret_string, "Action ");
```
Comment (ellychoi) [inline src/action_management/src/actionmanagement.c]: put in "Requested Action " just for continuity with the other strings

File: src/action_management/src/actionmanagement.c [hunk 8]
```diff
@@ -50,58 +51,103 @@ void action_type_free(action_type_t *a)
...
-}
-
-
-// KIND 3
-/* See actionmanagement.h */
-int do_item_item_action(game_t *g, action_type_t *a,
+    /* TODO: implement the rest of this function, using game state funcs
+     * Will perform the action if all checks pass (Sprint 4)
+     */
+    strcat(ret_string, "Requested movement ");
```
Comment (ellychoi) [inline src/action_management/src/actionmanagement.c]: should be "Requested action " as well

File: src/action_management/src/actionmanagement.c [hunk 9]
```diff
@@ -50,58 +51,103 @@ void action_type_free(action_type_t *a)
...
+        strcat(ret_string, "The action type provided is not the right kind");
+        return ret_string;
     }
-    printf("Performed action %s with %s on %s.",
-           a->c_name, direct->item_id, indirect->item_id);
-    return SUCCESS;
+    // checks if the action can be used on the item
+    int allowed = allowed_action(indirect, a->c_name);
+    if (allowed != SUCCESS) {
+        strcat(ret_string, "Action ");
```
Comment (ellychoi) [inline src/action_management/src/actionmanagement.c]: "Requested action "

File: src/action_management/src/actionmanagement.c (hunk 2 above)
Comment (ellychoi) [reply src/action_management/src/actionmanagement.c]: I don't have a preference, but I think it depends on how formal we want our engine to be when interacting with the game.

File: src/action_management/src/actionmanagement.c (hunk 1 above)
Comment (ellychoi) [reply src/action_management/src/actionmanagement.c]: yep, it should be!

Comment (MTUC08) [reply src/action_management/src/actionmanagement.c]: right my b

File: src/action_management/src/actionmanagement.c (hunk 2 above)
Comment (MTUC08) [reply src/action_management/src/actionmanagement.c]: I made it not the right kind mainly bc you have to malloc less space, but I can change it

File: src/action_management/src/actionmanagement.c (hunk 4 above)
Comment (MTUC08) [reply src/action_management/src/actionmanagement.c]: oh I think this is due to an error made when merging actions/dev into actions/sprint3refactoring

File: src/action_management/src/actionmanagement.c (hunk 6 above)
Comment (MTUC08) [reply src/action_management/src/actionmanagement.c]: No I mean like we don't know how long the string is going to be, but it should never be longer than 100 chars

Comment (MTUC08) [reply src/action_management/src/actionmanagement.c]: So this is the allocated memory

File: src/action_management/src/actionmanagement.c (hunk 7 above)
Comment (MTUC08) [reply src/action_management/src/actionmanagement.c]: I didn't do requested action b/c the action can't be requested

**Comment by lucyyutingli** [comment]
Referring back to the comments I made earlier, I still think "correct kind" is more formal/professional than "right kind". I hope I'm not being to nit-picky, what do you think @ellychoi?

**Comment by ellychoi** [comment]
@lucyyutingli @MTUC08 Yea that works for me I do think correct is more formal.

**Comment by MTUC08** [comment]
Closing Statement:
With this PR we refactored our action functions to return a string to CLI. The functions now check for whether the action type given is the correct kind, and whether it can be used on an item. If checks pass, the function will return a string to CLI saying the action has been requested. If any check fails, the function will return a tailored error message, also in the form of a string.

Note: After this PR to actions/dev, we need to wait on game-state to PR their updated items.h/c to master because our functions depend on a function they just added (issue #172), and will need that to be in place in order for the build to be successful.

Looking forward: 
In Sprint 4 we will refactor once again to add condition checking and effect distribution (#188 #190 #191 ), as well as switch the return format to an int for ease of error checking #245 in tests.

**Comment by mpcsh** [comment]
Issue Score: ✓+

Comments:
Great work!

**Pull request merged**